### PR TITLE
Keep original values when saving based on Map version

### PR DESCRIPTION
**Pull request by tomek-i**
Changed the saving routine to take into account what the current map version is
this allows to modify maps from terraria mobile (iOS), save them and load them up in the game again
without any issues.

### CODE REVIEW

**Comment by BinaryConstruct** [comment]
Thanks for the PR. These changes are pretty extensive so I will need to test a several scenarios. Alternatively, the previous release can be used for now.

File: TEditXna/Terraria/World.FileV2.cs [hunk 1]
```diff
@@ -21,8 +21,10 @@ private static void SaveV2(World world, BinaryWriter bw)
         {
             world.Validate();
 
-            // initialize tileframeimportance array if empty
-            if (TileFrameImportant == null || TileFrameImportant.Length < TileCount)
+            if (TileFrameImportant.Length < TileCount && TileFrameImportant.Length > 0)
```
Comment (BinaryConstruct) [inline TEditXna/Terraria/World.FileV2.cs:24]: If framed tiles change between versions there could be issues here. I don't know if this is something that can be fixed here without keeping track of framed tiles for every version.

File: TEditXna/Terraria/World.FileV2.cs [hunk 2]
```diff
@@ -310,7 +319,7 @@ public static int SaveNPCs(IEnumerable<NPC> npcs, BinaryWriter bw)
             foreach (NPC npc in npcs)
             {
                 bw.Write(true);
-                bw.Write(npc.SpriteId);
+                bw.Write(npc.Name);
```
Comment (BinaryConstruct) [inline TEditXna/Terraria/World.FileV2.cs:322]: Why was this change made? Is this an IOS save file format?

File: TEditXna/Terraria/World.FileV2.cs [hunk 3]
```diff
@@ -558,7 +606,7 @@ public static int SaveTileEntities(World w, BinaryWriter bw)
         {
             bw.Write(w.TileEntities.Count);
 
-            foreach(TileEntity tentity in w.TileEntities)
+            foreach (TileEntity tentity in w.TileEntities)
```
Comment (BinaryConstruct) [inline TEditXna/Terraria/World.FileV2.cs:609]: There are a lot of auto-formatting changes that make the changes harder to isolate. In the future it makes it easier to keep these as a separate commit or PR. I'm not saying that they shouldn't be done, just that it makes it more difficult to see what actually changed.

File: TEditXna/Terraria/World.FileV2.cs (hunk 1 above)
Comment (BinaryConstruct) [inline TEditXna/Terraria/World.FileV2.cs:24]: If framed tiles change between versions there could be issues here. I don't know if this is something that can be fixed here without keeping track of framed tiles for every version.

**Comment by BinaryConstruct** [comment]
If you can rebase and update this with the latest changes I will test and merge it. Could you also attach a world file to test with? Thanks.

**Comment by tomek-i** [comment]
An iOS world file?

**Comment by BinaryConstruct** [comment]
Yes, please.

**Comment by tomek-i** [comment]
Just downloaded terraria again on iOS 

Terraria IOS is still on v1.3.0.7.9 build 390

Test world is a new blank created world.

The other two worlds are more advanced with few buildings and chests and npcs etc.

By the way, where do I upload it here? Just in a new issue?


I did delete the local repo after few weeks of no activity in that project so I haven’t had the time to check it out again and migrate any of your other changes done for 1.4 ... I hope I can have a look / play on the weekend and then I can do the rebase

**Comment by tomek-i** [comment]
https://drive.google.com/file/d/1h2_RdiurIxxqFNwEOstRojbKqI-kBCGX/view?usp=drivesdk

**Comment by BinaryConstruct** [comment]
This has been merged. Thanks for the idea!

**Pull request closed without merging**